### PR TITLE
(PUP-8343) let rich data params through to resources

### PR DESCRIPTION
**Pull request by hlindberg**
Before this given rich-data (not all data types though) were transformed to String when given as arguments to classes and defines.
This was not good because it works against the purpose of type checking.

### CODE REVIEW

File: lib/puppet/pops/serialization/to_data_converter.rb [hunk 1]
```diff
@@ -188,17 +188,21 @@ def unknown_to_data(value)
...
 
     def unknown_to_string_with_warning(value)
-      str = value.to_s
+      str = unknown_to_string(value)
       serialization_issue(Issues::SERIALIZATION_UNKNOWN_CONVERTED_TO_STRING, :path => path_to_s, :klass => value.class, :value => str)
       str
     end
 
+    def unknown_to_string(value)
+      value.is_a?(Regexp) ? value.inspect : value.to_s
```
Comment (thallgren) [inline lib/puppet/pops/serialization/to_data_converter.rb]: Please use `PRegexpType.regexp_to_s` or  `PRegexpType.regexp_to_s_with_delimiters` whenever converting regexp to string. The output from the `inspect` method differ between ruby versions (has caused us a some grief until I cleaned it up everywhere).

Another option would be to use the StringFormatter with defaults (for all types of values, see line 249).

File: spec/integration/parser/compiler_spec.rb [hunk 2]
```diff
@@ -898,6 +898,24 @@ class { 'foo':  }
         end.to raise_error(/Class\[Foo\]: parameter 'x' expects an Integer value, got Undef/)
       end
 
+      it 'denies a regexp (rich data) argument given to class String parameter (even if later encoding of it is a string)' do
+        expect do
+          catalog = compile_to_catalog(<<-MANIFEST)
+            class foo(String $x) { }
+            class { 'foo':  x => /I am a regexp and I don't want to be a String/}
```
Comment (thallgren) [inline spec/integration/parser/compiler_spec.rb]: Perhaps you could use a regexp containing escaped forward slashes to ensure that escaping is consistent.

Comment (hlindberg) [reply spec/integration/parser/compiler_spec.rb]: We already have such a test elsewhere - it ensures that a regexp is formatted as expected.

Comment (thallgren) [reply spec/integration/parser/compiler_spec.rb]: For the particular case when the new unknown_to_string method is used?

**Comment by hlindberg** [comment]
Finally - travis goes green after 24h of having trouble running our tests... merging

**Pull request merged**